tensorflow/core/kernels/data/dataset_test_params.h

```diff
@@ -0,0 +1,208 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_CORE_KERNELS_DATA_DATASET_TEST_PARAMS_H_
+#define TENSORFLOW_CORE_KERNELS_DATA_DATASET_TEST_PARAMS_H_
+
+#include "tensorflow/core/framework/function.h"
+#include "tensorflow/core/framework/function_testlib.h"
+#include "tensorflow/core/framework/partial_tensor_shape.h"
+#include "tensorflow/core/framework/tensor_testutil.h"
+#include "tensorflow/core/kernels/data/batch_dataset_op.h"
+#include "tensorflow/core/kernels/data/dataset_utils.h"
+#include "tensorflow/core/kernels/data/map_dataset_op.h"
+#include "tensorflow/core/kernels/data/name_utils.h"
+#include "tensorflow/core/kernels/data/range_dataset_op.h"
+#include "tensorflow/core/lib/core/refcount.h"
+
+namespace tensorflow {
+namespace data {
+
+constexpr char kDefaultIteratorPrefix[] = "Iterator";
+
+// Creates a tensor with the specified dtype, shape, and value.
+template <typename T>
+static Tensor CreateTensor(const TensorShape& input_shape,
+                           const gtl::ArraySlice<T>& input_data) {
+  Tensor tensor(DataTypeToEnum<T>::value, input_shape);
+  test::FillValues<T>(&tensor, input_data);
+  return tensor;
+}
+
+// Creates a vector of tensors with the specified dtype, shape, and values.
+template <typename T>
+std::vector<Tensor> CreateTensors(
+    const TensorShape& shape, const std::vector<gtl::ArraySlice<T>>& values) {
+  std::vector<Tensor> result;
+  result.reserve(values.size());
+  for (auto& value : values) {
+    result.emplace_back(CreateTensor<T>(shape, value));
+  }
+  return result;
+}
+
+enum class DatasetParamsType {
+  Range,
+  Batch,
+  Map,
+};
+
+string ToString(DatasetParamsType type);
+
+class RangeDatasetParams;
+class BatchDatasetParams;
+class MapDatasetParams;
+
+class DatasetParams : public core::RefCounted {
+ public:
+  DatasetParams(DataTypeVector output_dtypes,
+                std::vector<PartialTensorShape> output_shapes, string node_name,
+                DatasetParamsType type);
+
+  ~DatasetParams() {}
+
+  // Returns the dataset inputs as a TensorValue vector.
+  virtual Status MakeInputs(gtl::InlinedVector<TensorValue, 4>* inputs) = 0;
+
+  // Checks if the tensor is a dataset variant tensor.
+  static bool IsDatasetTensor(const Tensor& tensor);
+
+  // Creates a BatchDatasetParams. The ownership is transferred to the caller.
+  BatchDatasetParams* Batch(int64 batch_size, bool drop_remainder,
```

Comment:
The ownership is transferred to the caller. That means the caller just needs to call `batch_dataset_params->UnRef()` or `core::ScopedUnref scoped_unref(batch_dataset_params)` to handle the deallocation for the whole pipeline.

Reply:
For example:
```C++
auto* range_dataset_params = new RangeDatasetParams(10, 0, -3);
auto* map_dataset_params = range_dataset_params->Batch(...)->Map(...);
core::ScopedUnref scoped_unref(map_dataset_params);
```
The allocated `range_dataset_params`, `batch_dataset_params`, and `map_dataset_params` will be automatically released.

Reply:
Having these as member methods only works for unary transformations. However, for dataset sources (e.g. `range` or `random`) or binary (e.g. `concatenate`) or n-nary transformation (e.g. `zip`) this will not work.

Instead of a member method, there should be a static factory that takes the input dataset parameters as a parameter:

```
static DatasetParams Range(...);
static DatasetParams Batch(DatasetParams params, ...);
static DatasetParams Zip(std::vector<DatasetParams> params, ...);
```

and then use it as follows:

```
auto params = DatasetParams::CreateMap(std::move(DatasetParams::CreateBatch(std::move(DatasetParams::CreateRange(10, 0, -3)), ...)), ...);
```

Reply:
@jsimsa Thanks for the prompt review! I have a few questions here.

>Instead of a member method, there should be a static factory that takes the input dataset parameters as a parameter

One potential issue for the static factory is that different static methods will have different input and return types as different DatasetParams have different attributes. For example:
```C++
static RangeDatasetParams Range(...);
static BatchDatasetParams Batch(T params, ...);
static ZipDatasetParams Zip(std::vector<DatasetParams> params, ...);
```
>`auto params = DatasetParams::Map(std::move(DatasetParams::Batch(std::move(DatasetParams::Range(10, 0, -3)), ...)), ...);`

The different types of input and return DatasetParams in each static method make it difficult to chain multiple dataset parameters together.

Reply:
Why can't they all return `DatasetParams` which the derived classes inherit from?

Reply:
I may misunderstand something. Here is an example:
```C++
static DatasetParams Range(int64 start, int64 stop, int64 step) {
  return RangeDatasetParams(start, stop, step);
}

DatasetParams range_dataset_params = Range(start, stop, step);
```
during the conversion, the values of `range_dataset_params.start`, `range_dataset_params.stop`, and `range_dataset_params.step` will be discarded.